apollo-router/src/plugins/README.md

```diff
@@ -0,0 +1,222 @@
+
+6. **Testing:**  
+   Write tests for your plugin to ensure it behaves as expected.
+
+7. **Configuration Example:**  
+   Provide an example of how to configure your plugin in the router's YAML configuration file.
+
+   Example YAML configuration:
+   ```yaml
+   plugins:
+     my_plugin:
+       enabled: true
+       timeout: 30s
+   ```
+
+By following these steps, you can create a custom plugin that extends Apollo Router's functionality.
+
+## Summary of Existing Plugins
+
+Below is a summary of the main plugins included in Apollo Router:
+
+### Authentication
+
+- **Purpose:** Verifies incoming requests using JWTs or other mechanisms. Supports extracting tokens from headers or cookies, and can be configured to allow or reject unauthenticated requests.
+- **Configurable:** Yes (JWT issuers, header names, error handling, etc.)
+
+### Authorization
+
+- **Purpose:** Enforces access control based on authentication status, scopes, or custom policies. Supports GraphQL directives like `@authenticated` and `@requiresScopes`.
+- **Configurable:** Yes (require authentication, enable/disable directives, error handling)
+
+### Cache
+
+- **Purpose:** Controls HTTP and entity caching, cache invalidation, and cache metrics. Supports fine-grained cache control headers and invalidation endpoints.
+- **Configurable:** Yes (cache control, invalidation rules, metrics)
+
+### Telemetry
+
+- **Purpose:** Collects and exports metrics and traces for observability. Integrates with OpenTelemetry, Prometheus, and Apollo Studio.
+- **Configurable:** Yes (metrics exporters, tracing, custom instruments)
+
+### Traffic Shaping
+
+- **Purpose:** Manages request flow with features like query deduplication, timeouts, compression, and rate limiting.
+- **Configurable:** Yes (enable/disable features, set limits, compression types)
+
+### Headers
+
+- **Purpose:** Allows insertion, removal, and propagation of HTTP headers at various stages. Supports static values, context-based, and body-based header manipulation.
+- **Configurable:** Yes (rules for each stage and target)
+
+### Coprocessor
+
+- **Purpose:** Offloads processing to external HTTP services at various pipeline stages (router, supergraph, execution, subgraph). Useful for custom logic or integrations.
+- **Configurable:** Yes (external URLs, which data to send, timeouts)
+
+### Content Negotiation
+
+- **Purpose:** Handles HTTP content negotiation using `Accept` and `Content-Type` headers. Ensures clients receive responses in supported formats.
+- **Configurable:** Minimal (mostly internal logic)
+
+### Subscription
+
+- **Purpose:** Adds support for GraphQL subscriptions, including callback and passthrough modes, deduplication, and connection management.
+- **Configurable:** Yes (modes, deduplication, limits)
+
+### CSRF Protection
+
+- **Purpose:** Protects against Cross-Site Request Forgery attacks by validating tokens in requests.
+- **Configurable:** Yes (token validation, error handling)
```

Comment:
The CSRF plugin doesn't do any token validation and doesn't have error configuration. Actually, it's the CORS plugin that prevents cross-site requests. The CSRF plugin specifically rejects requests that are exempt from CORS rules in the browser, and together this eliminates cross-site attacks.